experimental/QuadFormAndIsom/src/enumeration.jl

```diff
@@ -42,47 +48,45 @@ function _find_D(d::T, m::Int, p::IntegerUnion) where T <: IntegerUnion
   return D
 end
 
-# This is line 10 of Algorithm 1. We need the condition on the parity of
-# C since subgenera of an even genus are even too. r is the rank of
-# the subgenus, d its determinant, s and l the scale and level of C
-function _find_L(pG::Int, nG::Int, r::Int, d::RationalUnion, s::ZZRingElem, l::ZZRingElem, p::IntegerUnion, even = true; pos::Int=-1, neg::Int=-1)
-  def = ZZGenus[genus(integer_lattice(; gram=matrix(QQ, 0, 0, [])))]
-  if r == 0 && d == 1
-    return def
-  end
-  if pos >= 0
-    pos > pG && return def
-    if neg >= 0
-      pos + neg != r && return def
-    else
-      neg = r-pos
-    end
-    neg > nG && return def
-    gen = integer_genera((pos, neg), d; even)
-    filter!(G -> is_divisible_by(numerator(scale(G)), s), gen)
-    filter!(G -> is_divisible_by(p*l, numerator(level(G))), gen)
-  elseif neg >= 0
-    neg > nG && return def
-    pos = r-neg
-    pos > pG && return def
-    gen = integer_genera((pos, neg), d; even)
-    filter!(G -> is_divisible_by(numerator(scale(G)), s), gen)
-    filter!(G -> is_divisible_by(p*l, numerator(level(G))), gen)
-  else
-    gen = ZZGenus[]
-    for s1 in max(0, r-nG):min(pG, r)
-      s2 = r-s1
-      L = integer_genera((s1, s2), d; even)
-      filter!(G -> is_divisible_by(numerator(scale(G)), s), L)
-      filter!(G -> is_divisible_by(p*l, numerator(level(G))), L)
-      append!(gen, L)
-    end
+# This is line 10 of Algorithm 1 of [BH23]. We need the condition on the parity
+# of `C` since subgenera of an even genus are even too.
+# `r` is the rank of the subgenus, `d` its determinant, and `s` and `l` are
+# respectively the scale and the level of `C`.
+function _find_L(
+    pG::Int,
+    nG::Int,
+    r::Int,
+    d::RationalUnion,
+    s::ZZRingElem,
+    l::ZZRingElem,
+    p::Int,
+    even::Bool = true;
+    pos::AbstractVector{Int}=0:pG,
+    neg::AbstractVector{Int}=0:nG,
+  )
+  if r == 0 && d == 1 # In this case, we return the trivial genus as default
+    (!(0 in pos) || !(0 in neg)) && return ZZGenus[]
+    return ZZGenus[genus(integer_lattice(; gram=matrix(QQ, 0, 0, [])))]
+  end
+  gen = ZZGenus[]
+  Ip = intersect(max(0, r-nG):min(pG, r), pos) # Possible positive signature
+  In = intersect(max(0, r-pG):min(nG, r), neg) # Possible negative signature
+  for s1 in Ip
+    s2 = r-s1
+    !(s2 in In) && continue
+    # We enumerate all potential genera with the given invariants
+    append!(gen, integer_genera((s1, s2), d; min_scale=s, max_scale=p*l, even))
```

Comment:
Big improvement here: we enumerate less genus symbols by imposing the conditions about the scale and level directly into the `integer_genera` function